From 1c27bc8a2e5d2e0bf02fe68f5547087d51fb40d7 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Barnab=C3=A1s=20P=C5=91cze?=
 <barnabas.pocze@ideasonboard.com>
Date: Fri, 31 Jul 2026 12:02:22 +0200
Subject: [PATCH 35/68] ipa: libipa: agc: Store digital gain
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

While digital gain support is not implemented yet, start by storing
the suggested digital gain in the active state.

Signed-off-by: Barnabás Pőcze <barnabas.pocze@ideasonboard.com>
Reviewed-by: Jacopo Mondi <jacopo.mondi@ideasonboard.com>
Reviewed-by: Stefan Klug <stefan.klug@ideasonboard.com>
(cherry picked from commit c8662cb560f856e780e5b73a01f10d7b1b089522)
---
 src/ipa/libipa/agc.cpp | 7 ++++++-
 src/ipa/libipa/agc.h   | 1 +
 2 files changed, 7 insertions(+), 1 deletion(-)

diff --git a/src/ipa/libipa/agc.cpp b/src/ipa/libipa/agc.cpp
index ad19a1baa..46f75d621 100644
--- a/src/ipa/libipa/agc.cpp
+++ b/src/ipa/libipa/agc.cpp
@@ -131,6 +131,9 @@ namespace agc {
  * \var ActiveState::automatic.quantizationGain
  * \brief Automatic quantization gain multiplier
  *
+ * \var ActiveState::automatic.digitalGain
+ * \brief Automatic digital gain multiplier
+ *
  * \var ActiveState::automatic.yTarget
  * \brief Automatically determined luminance target
  *
@@ -414,6 +417,7 @@ int AgcAlgorithm::configure(agc::Session &session, agc::ActiveState &state,
 	state.automatic.gain = session.minAnalogueGain;
 	state.automatic.exposure = defExposure;
 	state.automatic.quantizationGain = 1;
+	state.automatic.digitalGain = 1;
 	state.automatic.yTarget = impl_.effectiveYTarget(0, 1);
 	state.manual.gain = state.automatic.gain;
 	state.manual.exposure = state.automatic.exposure;
@@ -735,13 +739,14 @@ void AgcAlgorithm::process(const agc::Session &session, agc::ActiveState &state,
 	state.automatic.exposure = newEv.exposureTime / lineDuration;
 	state.automatic.gain = newEv.analogueGain;
 	state.automatic.quantizationGain = newEv.quantizationGain;
+	state.automatic.digitalGain = newEv.digitalGain;
 	state.automatic.yTarget = newEv.yTarget;
 
 	LOG(Agc, Debug)
 		<< "exposure-time: " << newEv.exposureTime << ", "
 		<< "analogue-gain: " << state.automatic.gain << ", "
 		<< "quantization-gain: " << state.automatic.quantizationGain << ", "
-		<< "digital-gain: " << newEv.digitalGain;
+		<< "digital-gain: " << state.automatic.digitalGain;
 
 	/*
 	 * Expand the target frame duration so that we do not run faster than
diff --git a/src/ipa/libipa/agc.h b/src/ipa/libipa/agc.h
index 06600f020..7c5a7c351 100644
--- a/src/ipa/libipa/agc.h
+++ b/src/ipa/libipa/agc.h
@@ -58,6 +58,7 @@ struct ActiveState {
 		uint32_t exposure;
 		double gain;
 		double quantizationGain;
+		double digitalGain;
 		double yTarget;
 	} automatic;
 
